### ISSUE TITLE
Define instruction set

### ISSUE THREAD

**Issue by dirk** [WebAssembly/design#187]
Although the [AST semantics](https://github.com/WebAssembly/design/blob/master/AstSemantics.md) is a nice first start at defining how programs will be structured in WebAssembly, it seems as though a more complete definition of statements (functions and control flow) and expressions (instructions) would be useful for helping people understand the design of WebAssembly, what is possible with it, and where it fits on the spectrum of [abstraction](https://en.wikipedia.org/wiki/High-level_programming_language).

Perhaps the AST document could be reorganized into more logical sections: types (integers, floats, pointers/memory), statements, and expressions. Currently it's a slightly confusing hodgepodge of all three and more.

**Comment by dejlek** [comment]
Yes please. I was looking for ISA document for last half an hour! :)

**Comment by dirk** [comment]
Once #200 is merged then I'd be more than willing to start on this.

**Comment by titzer** [comment]
The AST semantics document does given an overview with roughly that organization, though it is informal. We are planning on developing a formal specification (with operational semantics!) that will describe the operations in more detail.

**Comment by sunfishcode** [comment]
#245 created a standalone Types section, which makes some progress here.

**Comment by sunfishcode** [comment]
#200 is merged, and as mentioned above, #245 makes several improvements. What do you think of the current text?

**Comment by dirk** [comment]
It's looking a lot better! Seems like you're starting to arrive at a fairly complete set of instructions/features. (Hopefully I can start contributing more, been swamped with job transition recently.) How do you feel about rethinking the organization of that document to start grouping specifications & instructions into even more clear functional sections (memory, pointers, functions, mathematics, and control flow)?

**Comment by sunfishcode** [comment]
I think we basically already have functional sections like what you describe, for Linear memory, Calls, Integer operations, Floating point operations, and so on. The thing that feels the most out of place to me is "Expressions with control flow". That hasn't been a high priority because it's an open question whether we actually want those expressions at all (#239 etc.) but if you want to reorganize it I wouldn't mind :-).

**Comment by sunfishcode** [comment]
What do you think of the current organization, ignoring expressions with control flow?

**Comment by sunfishcode** [comment]
Ping?

**Comment by sunfishcode** [comment]
Closing; expressions with control flow and control flow in general are being discussed elsewhere, and the remaining specific suggestions here are implemented. If the original reporter or anyone else has any more suggestions, please file a new issue!
